# Lab 05 - Pandas Part 2 - Andrew Badzioch

#### Attention:
- From this lab, you need to add observations for any cell output that adds a new piece of knowledge to this data processing project.

#### 1. Introduction: (complete this part)
- context (background): This is a dataset that provides information on the film industry.
- objective: For this lab, we will use different methods and attributes to extract information to gain a better understanding on how and when to use the pandas library in Python.
- data description: This dataset contains information about movies titles, their gross box office income, their rank from that income, production studio, and year released.

#### 2. Importing Libraries

In [1]:
# general libraries
import numpy as np
import pandas as pd

In [2]:
# to suppress warnings
import warnings
warnings.filterwarnings('ignore')

#### 3. Loading (Reading) the Dataset 

In [3]:
# let's load the dataset
df = pd.read_csv('movies.csv')

In [4]:
# create a copy of the file using a method()
movies = df.copy()

#### 4. Basic Data Exploration

In [5]:
# basic data exploration with head(): method to view the first 5 rows
# a negative correlation between the rank and gross (one goes up and the other goes down)
movies.head()

Rank                         Title       Studio       Gross  Year
0     1             Avengers: Endgame  Buena Vista  $2,796.30   2019
1     2                        Avatar          Fox  $2,789.70   2009
2     3                       Titanic    Paramount  $2,187.50   1997
3     4  Star Wars: The Force Awakens  Buena Vista  $2,068.20   2015
4     5        Avengers: Infinity War  Buena Vista  $2,048.40   2018

In [6]:
movies.tail()

Rank                     Title           Studio     Gross  Year
777   778                 Yogi Bear  Warner Brothers  $201.60   2010
778   779       Garfield: The Movie              Fox  $200.80   2004
779   780               Cats & Dogs  Warner Brothers  $200.70   2001
780   781  The Hunt for Red October        Paramount  $200.50   1990
781   782                  Valkyrie              MGM  $200.30   2008

#### Data Attributes: Complete this part
1. shape: shows how many rows and columns in the dataset
2. size: returns rows * columns
3. ndim: gives the number of dimensions of the dataset
4. dtype: give the types of data within the dataset
5. columns: returns the column names along with the data type within

In [7]:
movies.shape

(782, 5)

**Observations:**

In [8]:
movies.size

3910

In [9]:
movies.ndim

2

In [10]:
movies.dtypes # variables

Rank       int64
Title     object
Studio    object
Gross     object
Year       int64
dtype: object

In [11]:
movies.columns # variables

Index(['Rank', 'Title', 'Studio', 'Gross', 'Year'], dtype='object')

**Observations:**
- the shape of the dataset is 782 rows and 5 columns
- the size of the dataset is 3910 (rows(782) * columns(5))
- this is 2 dimensional, as it has both rows and columns

#### 5. Data Processing Functions

In [12]:
# Let's get more information about the dataset
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 782 entries, 0 to 781
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Rank    782 non-null    int64 
 1   Title   782 non-null    object
 2   Studio  782 non-null    object
 3   Gross   782 non-null    object
 4   Year    782 non-null    int64 
dtypes: int64(2), object(3)
memory usage: 30.7+ KB


**Observations:**
- class: the main information about the dataframe is
- RangeIndex: gives the total number of entries and the index range
- Data columns: the total number of columns along with their name, count, if there are any empty places, and data type 
- dtype: the type of data for every column (object=mixed values), (int64=integer)

In [13]:
# Let's get the statistical summary of the dataset
movies.describe() # 4 observations

Rank         Year
count  782.000000   782.000000
mean   391.500000  2006.620205
std    225.888247    10.026227
min      1.000000  1939.000000
25%    196.250000  2001.000000
50%    391.500000  2009.000000
75%    586.750000  2014.000000
max    782.000000  2019.000000

**Observations:**
- describe is used for numerical values not objects
- (std) Standard Deviation: a measure of how spread out the numbers are. A large deviation indicates that the data is spread out. A small deviation indicates that the data is clustered together.
- MINimum, MAXimum, and percentile placement within the dataset
- when the mean and median are the same, the dataset is well balanced

In [14]:
# Let's check for missing values in the dataset
movies.isnull().sum() # 2 observations

Rank      0
Title     0
Studio    0
Gross     0
Year      0
dtype: int64

**Observations:**
- .isnull() method returns True/False for empty/not empty information
- .sum() method returns the sum of empty information within the dataset

In [15]:
movies.head(3)

Rank              Title       Studio       Gross  Year
0     1  Avengers: Endgame  Buena Vista  $2,796.30   2019
1     2             Avatar          Fox  $2,789.70   2009
2     3            Titanic    Paramount  $2,187.50   1997

In [16]:
# Let's change the index title to No.# (number)
movies.index.name='No.'
movies.head(3)

Rank              Title       Studio       Gross  Year
No.                                                        
0       1  Avengers: Endgame  Buena Vista  $2,796.30   2019
1       2             Avatar          Fox  $2,789.70   2009
2       3            Titanic    Paramount  $2,187.50   1997

In [17]:
# Let's set Title column as the index
movies.set_index('Title', inplace=True)
movies.head(3)

Rank       Studio       Gross  Year
Title                                                 
Avengers: Endgame     1  Buena Vista  $2,796.30   2019
Avatar                2          Fox  $2,789.70   2009
Titanic               3    Paramount  $2,187.50   1997

##### Examining the unique values in each variable
- we use the value_counts() method to perform this process

In [18]:
# How many different studios are there in this dataset?
movies['Studio'].value_counts()

Studio
Warner Brothers           132
Buena Vista               125
Fox                       117
Universal                 109
Sony                       86
Paramount                  76
Dreamworks                 27
Lionsgate                  21
New Line                   16
MGM                        11
TriStar                    11
Miramax                    10
Weinstein                   6
Columbia                    5
WGUSA                       4
Polygram                    2
Orion                       2
SonR                        2
Dimension                   2
Vestron                     1
USA                         1
Lions                       1
Focus                       1
Rela.                       1
CL                          1
Pathe                       1
Artisan                     1
IFC                         1
GrtIndia                    1
RKO                         1
UTV                         1
FUN                         1
FR                          1
New

In [19]:
# What are the top five studios in the dataset
movies['Studio'].value_counts().head() # 4 observations

Studio
Warner Brothers    132
Buena Vista        125
Fox                117
Universal          109
Sony                86
Name: count, dtype: int64

**Observations:**
- from the movies DataFrame - movies
- use the index of ['Studio'] from the movies DataFrame
- sort the values in descending order with .value_counts() method
- limit the output to the first five rows with .head() method

In [20]:
# Let's see the movies from Universal Studio only 
movies[movies['Studio'] == 'Universal'] # double indexing (== is needed as it is not a variable)

Rank     Studio       Gross  Year
Title                                                            
Jurassic World                     6  Universal  $1,671.70   2015
Furious 7                          8  Universal  $1,516.00   2015
Jurassic World: Fallen Kingdom    13  Universal  $1,309.50   2018
The Fate of the Furious           17  Universal  $1,236.00   2017
Minions                           19  Universal  $1,159.40   2015
...                              ...        ...         ...   ...
The Break-Up                     763  Universal    $205.00   2006
Everest                          766  Universal    $203.40   2015
Patch Adams                      772  Universal    $202.30   1998
Kindergarten Cop                 775  Universal    $202.00   1990
Straight Outta Compton           776  Universal    $201.60   2015

[109 rows x 4 columns]

In [21]:
movies[movies['Studio'] == 'Universal'].head() # 4 observations

Rank     Studio       Gross  Year
Title                                                            
Jurassic World                     6  Universal  $1,671.70   2015
Furious 7                          8  Universal  $1,516.00   2015
Jurassic World: Fallen Kingdom    13  Universal  $1,309.50   2018
The Fate of the Furious           17  Universal  $1,236.00   2017
Minions                           19  Universal  $1,159.40   2015

**Observations:**
- 

#### Data Indexing and Filtration
1. iloc[] indexer: This indexer is implicit index (not clearly described)
    - when not defined, the iloc[] starts with 0
    - the i in iloc[] stands for integer
2. loc[] indexer: This indexer is explicit index (clearly stated)
    - the locator is the user defined name

In [22]:
# let's use the loc[] indexer
movies.loc['Forrest Gump'] # 2 observations

Rank            119
Studio    Paramount
Gross      $677.90 
Year           1994
Name: Forrest Gump, dtype: object

**Observations:**
- use .loc[] when you know the defined name
- will return all given column information about the row as well as data type

In [23]:
movies.loc['101 Dalmatians'] # 2 observations

Rank       Studio     Gross  Year
Title                                            
101 Dalmatians   425  Buena Vista  $320.70   1996
101 Dalmatians   708  Buena Vista  $215.90   1961

**Observations:**
- although pandas allows duplicates, try keeping labels unique to avoid issues
- unique labels accelerates the speed at which pandas extracts information

In [24]:
# let's use the iloc[] indexer (by its position)
movies.iloc[300] # 2 observations


Rank           301
Studio         Fox
Gross     $389.70 
Year          2016
Name: Independence Day: Resurgence, dtype: object

**Observations:**
- use .iloc[] to extract info by the numeric order, or its index postion
- remember that most programming languages, the index starts with zero

### Conclusion: 10 pionts

#### End of Lab 05 Part2